requests-for-comments/2024-05-28-monorepo.md

```diff
@@ -0,0 +1,46 @@
+# Request for comments: Monorepo - combine rust services and main posthog app
+
+## Problem statement
+*Who are we building for, what are their needs (include example cases), why is this important?*
+
+I've had a few conversations with everyone involved, and there hasn't been much pushback, so I'll very quickly summarise the problem, and then focus this RFC on how we get to what we want.
+
+Having things split up between hog-rs and posthog app make a couple things pretty problematic:
+1. Trying to develop locally with both, the capture-rs/decide-rs endpoints running locally, and the main posthog app running locally
+2. With shared data models (example: django app controlling flag CRUD and redis population, decide-rs using this data to evaluate flags) means it becomes much harder to keep things in sync, and making sure a data model update doesn't break the other service.
+3. It's harder to keep track of changes that need to be made everywhere / keeping code in sync across repos.
+4. We need to re-create similar scaffolding to make things work in the rust repo (example: GeoIP db downloading, webhooks data model setup, etc.)
+
+Now, I'm looking for alignment on the steps below, and whether this makes sense, and if I'm missing anything.
+
+## Design 
+*What are the key user experience and technical design decisions / trade-offs?*
+
+To make this work, here's what we need to do: (non-exhaustive, hoping you can add what I'm missing!)
+
+Effectively, nothing should change for anyone developing on `hog-rs` currently, and people developing on main `posthog` repo should get more options to hook things up.
+
+1. [ ] Introduce a `rust/` top-level folder that has all rust services
```

Comment:
If we go with a rust top level it should make the whole "only run the CI for this code" quite straightforward.

Reply:
Also one thought - you could do this in the `hog-rs` repo first to test out that it works as expected.

Reply:
For a True Monorepo™ it would also be great to put everything which is not `rust/`/`plugin-server/`/`plugin-transpiler/` under `web/` – since it doesn't make sense that the plugin server and such are _under_ the Django server in the hierarchy. Architecturally, they're all on the same level. (This is scope creep)